HopsanGUI/Widgets/TextEditorWidget.cpp

```diff
@@ -0,0 +1,682 @@
+            QString text = textCursor().selection().toPlainText();
+            text.replace("\n# ", "\n");
+            if(text.startsWith("# ")) {
+                text.remove(0,2);
+            }
+            int pos = textCursor().position();
+            int stopPos = textCursor().anchor();
+            QTextCursor c;
+            c = textCursor();
+            textCursor().removeSelectedText();
+            insertPlainText(text);
+
+            c.setPosition(std::min(pos,stopPos));
+            setTextCursor(c);
+
+            textCursor().endEditBlock();
+        }
+    }
+    else
+    {
+        QPlainTextEdit::keyPressEvent(event);
+
+        updateAutoCompleteList();
+
+        QTextCursor tc = textCursor();
+        tc.select(QTextCursor::WordUnderCursor);
+        QString prefix = tc.selectedText();
+        if((event->key() != Qt::Key_Space || !event->modifiers().testFlag(Qt::ControlModifier)) && prefix.isEmpty())
+            return;
+
+        if (prefix != mpCompleter->completionPrefix()) {
+            mpCompleter->setCompletionPrefix(prefix);
+            mpCompleter->popup()->setCurrentIndex(mpCompleter->completionModel()->index(0, 0));
+        }
+        QRect cr = cursorRect();
+        cr.setWidth(mpCompleter->popup()->sizeHintForColumn(0)
+                    + mpCompleter->popup()->verticalScrollBar()->sizeHint().width());
+        mpCompleter->complete(cr);
+    }
+}
+
+
+void TextEditor::wheelEvent(QWheelEvent* event)
+{
+#if QT_VERSION >= 0x050000  //zoomIn() and zoomOut() not available in Qt4
+   if ((event->modifiers() == Qt::ControlModifier) && (event->delta() > 0))
+       zoomIn(2);
+   else if ((event->modifiers() == Qt::ControlModifier) && (event->delta() < 0))
+       zoomOut(2);
+   else
+       QPlainTextEdit::wheelEvent(event);
+#else
+   QPlainTextEdit::wheelEvent(event);
+#endif
+}
+
+
+//! @brief Updates list of auto complete words
+void TextEditor::updateAutoCompleteList()
+{
+    if(mLanguage == HighlighterTypeEnum::Hcom) {
+        mpCompleter->setModel(new QStringListModel(gpTerminalWidget->mpHandler->getAutoCompleteWords(), mpCompleter));
+    }
+    else if(mLanguage == HighlighterTypeEnum::Cpp) {
+        QStringList lines = toPlainText().split("\n");
+
+        QStringList dataTypes = QStringList() << "double" << "int" << "SecondOrderTransferFunction" << "FirstOrderTransferFunction" << "Port";
+        QStringList functions = QStringList() << "addInputVariable" << "addOutputVariable" << "addConstant" << "addPowerPort" << "getSafeNodeDataPtr";
+
+        int bracketCounter=-1;
+
+        QStringList variables;
+        foreach(const QString &line, lines)
+        {
+            if(line.simplified().startsWith("class "))
+                bracketCounter = 0;
+
+            bracketCounter += line.count("{");
+            bracketCounter -= line.count("}");
+            if(bracketCounter != 1)
+                continue;
+
+            if(line.contains("()")) //Ignore functions
```

Comment:
What if function has arguments

Reply:
Not sure actually, but it does not seem to be a problem. Better leave this as it is I guess. This line may not even be necessary.